File: project/drivers/driver/go/keepalive.go
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

package scratchbird

import (
	"context"
	"sync"
	"time"
)

// KeepaliveConfig configures connection keepalive behavior
type KeepaliveConfig struct {
	// How often to check idle connections (default: 2 minutes)
	Interval time.Duration
	// Maximum time a connection can be idle before validation (default: 10 minutes)
	MaxIdleBeforeCheck time.Duration
	// Timeout for validation query (default: 5 seconds)
	ValidationTimeout time.Duration
}

// DefaultKeepaliveConfig returns default keepalive configuration
func DefaultKeepaliveConfig() KeepaliveConfig {
	return KeepaliveConfig{
		Interval:           2 * time.Minute,
		MaxIdleBeforeCheck: 10 * time.Minute,
		ValidationTimeout:  5 * time.Second,
	}
}

// KeepaliveTracker tracks activity time for connection keepalive
type KeepaliveTracker struct {
	lastActivity sync.RWMutex
	lastTime     time.Time
	config       KeepaliveConfig
}

// NewKeepaliveTracker creates a new keepalive tracker
func NewKeepaliveTracker(config KeepaliveConfig) *KeepaliveTracker {
	return &KeepaliveTracker{
		lastTime: time.Now(),
		config:   config,
	}
}

// MarkActive marks the connection as active
func (k *KeepaliveTracker) MarkActive() {
	k.lastActivity.Lock()
	defer k.lastActivity.Unlock()
	k.lastTime = time.Now()
}

// NeedsValidation checks if connection needs validation
func (k *KeepaliveTracker) NeedsValidation() bool {
	k.lastActivity.RLock()
	defer k.lastActivity.RUnlock()
	return time.Since(k.lastTime) > k.config.MaxIdleBeforeCheck
}

// IdleDuration returns time since last activity
func (k *KeepaliveTracker) IdleDuration() time.Duration {
	k.lastActivity.RLock()
	defer k.lastActivity.RUnlock()
	return time.Since(k.lastTime)
}

// KeepaliveManager manages keepalive for a pool of connections
type KeepaliveManager struct {
	config   KeepaliveConfig
	ctx      context.Context
	cancel   context.CancelFunc
	checkers map[string]*KeepaliveChecker
	mu       sync.RWMutex
}

// KeepaliveChecker checks a single connection
type KeepaliveChecker struct {
	connID  string
	tracker *KeepaliveTracker
	pinger  func() error
}

// NewKeepaliveManager creates a new keepalive manager
func NewKeepaliveManager(config KeepaliveConfig) *KeepaliveManager {
	ctx, cancel := context.WithCancel(context.Background())
	
	km := &KeepaliveManager{
		config:   config,
		ctx:      ctx,
		cancel:   cancel,
		checkers: make(map[string]*KeepaliveChecker),
	}
	
	// Start background check loop
	go km.checkLoop()
	
	return km
}

// Register registers a connection for keepalive monitoring
func (km *KeepaliveManager) Register(connID string, pinger func() error) *KeepaliveTracker {
	tracker := NewKeepaliveTracker(km.config)
	
	checker := &KeepaliveChecker{
		connID:  connID,
		tracker: tracker,
		pinger:  pinger,
	}
	
	km.mu.Lock()
	km.checkers[connID] = checker
	km.mu.Unlock()
	
	return tracker
}

// Unregister removes a connection from keepalive monitoring
func (km *KeepaliveManager) Unregister(connID string) {
	km.mu.Lock()
	delete(km.checkers, connID)
	km.mu.Unlock()
}

// checkLoop runs periodic keepalive checks
func (km *KeepaliveManager) checkLoop() {
	ticker := time.NewTicker(km.config.Interval)
	defer ticker.Stop()
	
	for {
		select {
		case <-km.ctx.Done():
			return
		case <-ticker.C:
			km.checkConnections()
		}
	}
}

// checkConnections validates idle connections
func (km *KeepaliveManager) checkConnections() {
	km.mu.RLock()
	checkers := make([]*KeepaliveChecker, 0, len(km.checkers))
	for _, c := range km.checkers {
		checkers = append(checkers, c)
	}
	km.mu.RUnlock()
	
	for _, checker := range checkers {
		if checker.tracker.NeedsValidation() {
			ctx, cancel := context.WithTimeout(km.ctx, km.config.ValidationTimeout)
			
			// Run ping in goroutine to respect timeout
			done := make(chan error, 1)
			go func() {
				done <- checker.pinger()
			}()
			
			select {
			case err := <-done:
				if err != nil {
					// Connection is dead - mark as needing replacement
					// The pool will handle this on next checkout
				}
				checker.tracker.MarkActive() // Reset timer even on failure
			case <-ctx.Done():
				// Timeout - connection may be slow or dead
			}
			
			cancel()
		}
	}
}

// Stop stops the keepalive manager
func (km *KeepaliveManager) Stop() {
	km.cancel()
}

File: project/drivers/driver/rust/src/keepalive.rs
// Copyright (c) 2026 ScratchBird Software Inc.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
//
// SPDX-License-Identifier: MPL-2.0

// Keepalive Module for Rust Driver
// Prevents connection timeouts by periodically validating idle connections

use std::time::{Duration, Instant};
use tokio::sync::RwLock;
use tokio::time::interval;

/// Configuration for connection keepalive
#[derive(Debug, Clone, Copy)]
pub struct KeepaliveConfig {
    /// How often to check idle connections (default: 2 minutes)
    pub interval: Duration,
    /// Maximum time a connection can be idle before validation (default: 10 minutes)
    pub max_idle_before_check: Duration,
    /// Timeout for validation query (default: 5 seconds)
    pub validation_timeout: Duration,
}

impl Default for KeepaliveConfig {
    fn default() -> Self {
        Self {
            interval: Duration::from_secs(120),              // 2 minutes
            max_idle_before_check: Duration::from_secs(600), // 10 minutes
            validation_timeout: Duration::from_secs(5),
        }
    }
}

/// Tracks last activity time for connection keepalive
#[derive(Debug)]
pub struct KeepaliveTracker {
    last_activity: RwLock<Instant>,
    config: KeepaliveConfig,
}

impl KeepaliveTracker {
    pub fn new(config: KeepaliveConfig) -> Self {
        Self {
            last_activity: RwLock::new(Instant::now()),
            config,
        }
    }

    /// Mark connection as active
    pub async fn mark_active(&self) {
        let mut last = self.last_activity.write().await;
        *last = Instant::now();
    }

    /// Check if connection needs validation
    pub async fn needs_validation(&self) -> bool {
        let last = self.last_activity.read().await;
        last.elapsed() > self.config.max_idle_before_check
    }

    /// Get time since last activity
    pub async fn idle_duration(&self) -> Duration {
        let last = self.last_activity.read().await;
        last.elapsed()
    }
}

/// Background keepalive task for connection pools
pub struct KeepaliveTask {
    config: KeepaliveConfig,
    shutdown_tx: tokio::sync::watch::Sender<bool>,
}

impl KeepaliveTask {
    pub fn new(config: KeepaliveConfig) -> (Self, tokio::sync::watch::Receiver<bool>) {
        let (shutdown_tx, shutdown_rx) = tokio::sync::watch::channel(false);
        (
            Self {
                config,
                shutdown_tx,
            },
            shutdown_rx,
        )
    }

    /// Start the keepalive background task
    pub fn spawn<F, Fut>(&self, mut shutdown_rx: tokio::sync::watch::Receiver<bool>, check_fn: F)
    where
        F: Fn() -> Fut + Send + Sync + 'static,
        Fut: std::future::Future<Output = ()> + Send + 'static,
    {
        let interval_duration = self.config.interval;

        tokio::spawn(async move {
            let mut ticker = interval(interval_duration);

            loop {
                tokio::select! {
                    _ = ticker.tick() => {
                        check_fn().await;
                    }
                    _ = shutdown_rx.changed() => {
                        if *shutdown_rx.borrow() {
                            break;
                        }
                    }
                }
            }
        });
    }

    /// Shutdown the keepalive task
    pub fn shutdown(&self) {
        let _ = self.shutdown_tx.send(true);
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_keepalive_tracker() {
        let config = KeepaliveConfig {
            interval: Duration::from_secs(1),
            max_idle_before_check: Duration::from_millis(100),
            validation_timeout: Duration::from_secs(1),
        };

        let tracker = KeepaliveTracker::new(config);

        // Initially should not need validation
        assert!(!tracker.needs_validation().await);

        // Wait for idle threshold
        tokio::time::sleep(Duration::from_millis(150)).await;

        // Now should need validation
        assert!(tracker.needs_validation().await);

        // Mark active
        tracker.mark_active().await;

        // Should not need validation anymore
        assert!(!tracker.needs_validation().await);
    }
}
